### ISSUE TITLE
convert sparse edge_index of some graphs to mini-batch format

### ISSUE THREAD

**Issue by A-student-without-emotion** [rusty1s/pytorch_geometric#2766]
hey, I want to build a dynamic graph model that update edge_index dynamically.
now I have some graphs(eg 32 graphs) with PyG edge_index format, and I want to convert the 32 graphs to mini-batch format just like DataLoader
eg convert graph1 [[0,1...],[2,0...]]   graph2 [[0,2...],[1,2...]].......graph32   to mini-batch edge_index format [[...],[.....]]
could give me a way to achieve it?

**Comment by rusty1s** [comment]
Our `DataLoader` does not support mini-batching of nested arrays or dictionaries yet, so the best way I can think of is that you save your final graph in `edge_index`, and maintain a mask tensor of shape `[num_timestamps, num_edges]`, which you can use to filter `edge_index` for each timestamp. WDYT?

**Comment by A-student-without-emotion** [comment]
> Our `DataLoader` does not support mini-batching of nested arrays or dictionaries yet, so the best way I can think of is that you save your final graph in `edge_index`, and maintain a mask tensor of shape `[num_timestamps, num_edges]`, which you can use to filter `edge_index` for each timestamp. WDYT?

Thank you very much for your advice!! but I have a question
At the beginning, I want to update the edge_index in the model dynamically, but I find that even if I create the same edge_index format like DataLoader,the model does't work(accuracy of test set does not rise, but rise without self.update). I think there are some problems here, but I'm a beginner. I don't know what causes the model not to work. 
could tell me why?
```
def  forward(self, x, edge_index, batch, some_features):
         edge_index = self.update(some_features)
         ............
```

**Comment by rusty1s** [comment]
I'm not entirely sure what `update` is trying to do. Can you elaborate?

**Comment by A-student-without-emotion** [comment]
> I'm not entirely sure what `update` is trying to do. Can you elaborate?

Thank you for your reply！
`some_features` is a square matrix with different values, and `update` use a learnable parameter to filter values in `some_features`, and I use the position(row, col) of filtered values to form a new edge_index mini-batch format

**Comment by rusty1s** [comment]
Ok, can you share your code with me?

**Comment by A-student-without-emotion** [comment]
> Ok, can you share your code with me?

this is my test model
fc_matrix is added in Data, and I use `to_dense_batch` to restore the matrix.

```
class learnable_graph(torch.nn.Module):
    def __init__(self, num_node_feature, hidden_channels, max_site_value, target_label_num=2):
        super(learnable_graph, self).__init__()
        torch.manual_seed(12345)
        self.max_site_value = max_site_value
        self.conv1 = GraphConv(num_node_feature, hidden_channels)
        self.conv2 = GraphConv(hidden_channels, hidden_channels)
        self.fc = Linear(hidden_channels, target_label_num)
        self.sigmoid = Sigmoid()

        # learnable filter
        self.r = torch.nn.Parameter(torch.FloatTensor(1), requires_grad=True)

    def update_edge_index(self, fc_matrix):
        # return edge_index of mini-batch format
        edge_index = []
        i = 0
        for each_matrix in fc_matrix:
            # the range of fc_matrix is [-1, 1] so sigmoid make sure the range of r to [-1,1]
            edge_index.append(
                torch.nonzero(each_matrix > self.sigmoid(self.r)).t().contiguous() + 160 * i)  # each graph has 160 nodes
            i += 1

        return torch.cat(edge_index, dim=1)

    def forward(self, x, edge_index, batch, fc_matrix):
  
        edge_index = self.update_edge_index(to_dense_batch(fc_matrix, batch)[0])
        print(edge_index)
        x = self.conv1(x, edge_index)
        x = self.conv2(x, edge_index)
        x = global_mean_pool(x, batch)
        x = F.dropout(x, p=0.5, training=self.training)
        x = self.fc(x)
        return x
```

**Comment by rusty1s** [comment]
I see. Note that there aren't any gradients going through `update_edge_index` as they `nonzero` call will destroy them. You need to also keep the probability of an edge (i.e., its `edge_weight`) in order to make `fc_matrix` trainable):
```python
def update_edge_index(self, fc_matrix):
    edge_weight = []
    for ...
        edge_weight.append(each_matrix[edge_index[0], edge_index[1]])

    return torch.cat(edge_index, dim=1), torch.cat(edge_weight, dim=0)

def forward(...):
    edge_index, edge_weight = self.update_edge_index(...)
    x = self.conv1(x, edge_index, edge_weight)
```

**Comment by A-student-without-emotion** [comment]
> I see. Note that there aren't any gradients going through `update_edge_index` as they `nonzero` call will destroy them. You need to also keep the probability of an edge (i.e., its `edge_weight`) in order to make `fc_matrix` trainable):
> 
> ```python
> def update_edge_index(self, fc_matrix):
>     edge_weight = []
>     for ...
>         edge_weight.append(each_matrix[edge_index[0], edge_index[1]])
> 
>     return torch.cat(edge_index, dim=1), torch.cat(edge_weight, dim=0)
> 
> def forward(...):
>     edge_index, edge_weight = self.update_edge_index(...)
>     x = self.conv1(x, edge_index, edge_weight)
> ```

thanks!!  
The big problem is tha there aren't any gradients going through update_edge_index beacuse of  the `nonzero`, but `nonzero` is very convenient to fast get the position of all filtered values.
I would appreciate it if you could give me some advice to solve this problem (keeping gradients going through update_edge_index)

**Comment by rusty1s** [comment]
Yes, I did so above. You need to use the resulting `edge_index` to gather the values from `fc_matrix`.
